### PR TITLE
MIDI clock sync mechanism

### PR DESCRIPTION
**Pull request by awhitty**
Usage is something like this:

``` java
LXMidiInput clock;

// in setup():
clock = LXMidiSystem.matchInput(lx, "IAC Traktor");
clock.enableMidiClock();
```

### CODE REVIEW

File: src/heronarts/lx/midi/LXMidiInput.java [hunk 1]
```diff
@@ -118,6 +141,25 @@ public void send(MidiMessage midiMessage, long timeStamp) {
...
 
+  private void handleClockSignal(ShortMessage sm) {
+    Tempo tempo = this.midiEngine.lx.tempo;
+    switch (sm.getStatus()) {
+    case ShortMessage.TIMING_CLOCK:
+      if (this.midiClockCounter == 0) {
+        tempo.tap(false);
+      }
+      this.midiClockCounter += 1;
+      this.midiClockCounter %= 24;
```
Comment (mcslee) [inline src/heronarts/lx/midi/LXMidiInput.java]: And up top you should:
private final static int MIDI_PULSES_PER_QUARTER_NOTE = 24;
this.midiClockCounter = (this.midiClockCounter + 1) % MIDI_PULSES_PER_QUARTER_NOTE;

File: src/heronarts/lx/midi/LXMidiInput.java [hunk 2]
```diff
@@ -118,6 +141,25 @@ public void send(MidiMessage midiMessage, long timeStamp) {
     }
   }
 
+  private void handleClockSignal(ShortMessage sm) {
+    Tempo tempo = this.midiEngine.lx.tempo;
```
Comment (mcslee) [inline src/heronarts/lx/midi/LXMidiInput.java]: This is a style nitpick, but I think it's more readable not to store this in a local variable. Just reference it directly in the switch blocks.

File: src/heronarts/lx/midi/LXMidiInput.java [hunk 3]
```diff
@@ -118,6 +141,25 @@ public void send(MidiMessage midiMessage, long timeStamp) {
...
+    switch (sm.getStatus()) {
+    case ShortMessage.TIMING_CLOCK:
+      if (this.midiClockCounter == 0) {
+        tempo.tap(false);
+      }
+      this.midiClockCounter += 1;
+      this.midiClockCounter %= 24;
+      break;
+    case ShortMessage.START:
+      tempo.trigger();
```
Comment (mcslee) [inline src/heronarts/lx/midi/LXMidiInput.java]: I'm wondering if you want to also reset this.midiClickCounter to 0 here? There could be cases where midi clock stops being sent, and then a new start() signal is sent without there necessarily ever having been a STOP signal sent.

File: src/heronarts/lx/Tempo.java [hunk 4]
```diff
@@ -203,6 +203,25 @@ public void tap() {
...
+  public void tap(boolean smoothingEnabled) {
+    if (!smoothingEnabled) {
+      long now = System.currentTimeMillis();
+      this.firstTap = lastTap;
+      this.lastTap = now;
+
+      double beatPeriod = this.lastTap - this.firstTap;
+      setBpm(MINUTE / beatPeriod);
+    } else {
+      tap();
```
Comment (mcslee) [inline src/heronarts/lx/Tempo.java]: This is a confusing construction. I would modify this such that you have:

public void tap() {
  tap(true);
}

Then, in this method, show the logic for both the smoothed version as well as the non-smoothed version.

**Comment by awhitty** [comment]
Finally got around to wrapping this up :)

**Comment by awhitty** [comment]
Closing this to clean up my `/pulls`

**Pull request closed without merging**